File: app/src/main/java/com/livewallpaper/parallaxdemo/ParallaxWallpaperService.java
package com.livewallpaper.parallaxdemo;

import android.graphics.Canvas;
import android.os.Handler;
import android.os.HandlerThread;
import android.service.wallpaper.WallpaperService;
import android.view.SurfaceHolder;

public final class ParallaxWallpaperService extends WallpaperService {
    @Override
    public Engine onCreateEngine() {
        return new ParallaxEngine();
    }

    private final class ParallaxEngine extends Engine implements TiltSensor.Listener {
        private final HandlerThread renderThread = new HandlerThread("parallax-wallpaper-render");
        private Handler renderHandler;
        private ParallaxScene scene;
        private TiltSensor tiltSensor;
        private volatile float targetX;
        private volatile float targetY;
        private float currentX;
        private float currentY;
        private boolean visible;
        private boolean surfaceReady;

        private final Runnable drawFrame = new Runnable() {
            @Override
            public void run() {
                draw();
                if (visible && surfaceReady) {
                    renderHandler.postDelayed(this, 20L);
                }
            }
        };

        @Override
        public void onCreate(SurfaceHolder surfaceHolder) {
            super.onCreate(surfaceHolder);
            setOffsetNotificationsEnabled(false);
            renderThread.start();
            renderHandler = new Handler(renderThread.getLooper());
            renderHandler.post(() -> scene = new ParallaxScene(ParallaxWallpaperService.this));
            tiltSensor = new TiltSensor(ParallaxWallpaperService.this, this);
        }

        @Override
        public void onSurfaceCreated(SurfaceHolder holder) {
            super.onSurfaceCreated(holder);
            surfaceReady = true;
            scheduleDraw();
        }

        @Override
        public void onSurfaceChanged(SurfaceHolder holder, int format, int width, int height) {
            super.onSurfaceChanged(holder, format, width, height);
            scheduleDraw();
        }

        @Override
        public void onSurfaceDestroyed(SurfaceHolder holder) {
            surfaceReady = false;
            if (renderHandler != null) renderHandler.removeCallbacks(drawFrame);
            super.onSurfaceDestroyed(holder);
        }

        @Override
        public void onVisibilityChanged(boolean isVisible) {
            visible = isVisible;
            if (isVisible) {
                tiltSensor.start(renderHandler);
                scheduleDraw();
            } else {
                tiltSensor.stop();
                if (renderHandler != null) renderHandler.removeCallbacks(drawFrame);
            }
        }

        @Override
        public void onTilt(float x, float y) {
            targetX = x;
            targetY = y;
        }

        @Override
        public void onDestroy() {
            visible = false;
            surfaceReady = false;
            tiltSensor.stop();
            if (renderHandler != null) {
                renderHandler.removeCallbacksAndMessages(null);
                renderHandler.post(() -> {
                    if (scene != null) scene.release();
                });
            }
            renderThread.quitSafely();
            super.onDestroy();
        }

        private void scheduleDraw() {
            if (renderHandler == null || !visible || !surfaceReady) return;
            renderHandler.removeCallbacks(drawFrame);
            renderHandler.post(drawFrame);
        }

        private void draw() {
            if (scene == null || !surfaceReady) return;
            currentX += (targetX - currentX) * 0.24f;
            currentY += (targetY - currentY) * 0.24f;

            SurfaceHolder holder = getSurfaceHolder();
            Canvas canvas = null;
            try {
                canvas = holder.lockHardwareCanvas();
                if (canvas != null) scene.draw(canvas, currentX, currentY);
            } catch (RuntimeException ignored) {
                // Surface can disappear between visibility and draw callbacks.
            } finally {
                if (canvas != null) holder.unlockCanvasAndPost(canvas);
            }
        }
    }
}

File: packages/wallpaper-android/android/src/main/kotlin/com/qingjing/wallpaper_android/playback/ParallaxWallpaperService.kt
package com.qingjing.wallpaper_android.playback

import android.content.pm.ApplicationInfo
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.service.wallpaper.WallpaperService
import android.util.Log
import android.view.SurfaceHolder

class ParallaxWallpaperService : WallpaperService() {
    companion object {
        private val engines = mutableMapOf<Int,() -> Map<String,Any?>>()
        @Synchronized fun debugStates() = engines.values.map { it() }
        @Synchronized private fun register(id: Int,state: (() -> Map<String,Any?>)?) { if (state == null) engines.remove(id) else engines[id] = state }
    }
    override fun onCreateEngine(): Engine = ParallaxEngine()
    inner class ParallaxEngine : Engine() {
        private val main = Handler(Looper.getMainLooper())
        private var surfaceReady = false; private var dead = false; private var token: String? = null
        private var lastSelection: Pair<String,String?>? = null
        private var lastState: Map<String,Any?>? = null
        private lateinit var renderer: ParallaxSurfaceRenderer
        private fun state() = renderer.state()+mapOf("type" to "LAYER_PARALLAX","preview" to isPreview,"visible" to isVisible)
        private fun snapshot() {
            if (dead) return
            val value = state().filterKeys { it !in setOf("frames","motionObserved") }
            if (lastState != value && applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE != 0) Log.d("QJPlayback","elapsed=${SystemClock.elapsedRealtime()} engine=${hashCode()} $value")
            lastState = value
        }
        override fun onCreate(holder: SurfaceHolder) {
            super.onCreate(holder); setOffsetNotificationsEnabled(false)
            renderer = ParallaxSurfaceRenderer(this@ParallaxWallpaperService,holder,
                ready = { id,_ -> main.post { if (!dead && isVisible && surfaceReady && lastSelection?.first == id) { token?.let { LiveSelection.ready(this@ParallaxWallpaperService,it) }; snapshot() } } },
                failed = { main.post { snapshot() } },changed = { main.post { snapshot() } })
            register(hashCode()) { state() }
        }
        private val checkSelection = object : Runnable {
            override fun run() { if (!dead && isVisible && surfaceReady) { load(); main.postDelayed(this,1000) } }
        }
        private fun selection(): Pair<String,String?>? = if (isPreview) LiveSelection.pending(this@ParallaxWallpaperService,"LAYER_PARALLAX")
            else LiveSelection.current(this@ParallaxWallpaperService,"LAYER_PARALLAX")?.let { it to null }
        private fun load() {
            val selected = selection()
            if (selected == lastSelection) return
            lastSelection = selected; token = selected?.second
            if (selected == null) renderer.stop() else renderer.start(selected.first)
        }
        override fun onVisibilityChanged(visible: Boolean) {
            main.removeCallbacks(checkSelection); lastSelection = null
            if (visible && surfaceReady) { load(); main.postDelayed(checkSelection,1000) }
            else { token = null; renderer.stop() }
        }
        override fun onSurfaceCreated(holder: SurfaceHolder) { super.onSurfaceCreated(holder); surfaceReady = true; if (isVisible) { load(); main.removeCallbacks(checkSelection); main.postDelayed(checkSelection,1000) } }
        override fun onSurfaceChanged(holder: SurfaceHolder,format: Int,width: Int,height: Int) { renderer.redraw() }
        override fun onSurfaceDestroyed(holder: SurfaceHolder) { surfaceReady = false; lastSelection = null; token = null; renderer.stop(); main.removeCallbacks(checkSelection); super.onSurfaceDestroyed(holder) }
        override fun onDestroy() { dead = true; main.removeCallbacks(checkSelection); renderer.close(); register(hashCode(),null); super.onDestroy() }
    }
}
